### PR TITLE
[SM-581] User access removal warnings

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## Type of change

```
- [X] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The purpose of this PR is to provide the user a warning dialog when access policy removal will affect their access.

Needs supporting server PR https://github.com/bitwarden/server/pull/2762

## Code changes

- apps/web/src/locales/en/messages.json

Text for warning dialog.

- bitwarden_license/bit-web/src/app/secrets-manager/models/view/access-policy.view.ts
bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/models/responses/access-policy.response.ts

Adding new property from the server for detecting if a policy targets the current user.

- bitwarden_license/bit-web/src/app/secrets-manager/projects/project/project-people.component.html
bitwarden_license/bit-web/src/app/secrets-manager/projects/project/project-people.component.ts

Move to emitters and add in logic to launch warning dialog if access changes would remove the current user's access.

- bitwarden_license/bit-web/src/app/secrets-manager/projects/project/project-service-accounts.component.html
bitwarden_license/bit-web/src/app/secrets-manager/projects/project/project-service-accounts.component.ts

Move to emitters.

- bitwarden_license/bit-web/src/app/secrets-manager/projects/projects/projects.component.ts

Change logic, so the project list is refreshed on access lose.

- bitwarden_license/bit-web/src/app/secrets-manager/service-accounts/people/service-account-people.component.html
bitwarden_license/bit-web/src/app/secrets-manager/service-accounts/people/service-account-people.component.ts

Move to emitters and add in logic to launch warning dialog if access changes would remove the current user's access.

- bitwarden_license/bit-web/src/app/secrets-manager/service-accounts/projects/service-account-projects.component.html
bitwarden_license/bit-web/src/app/secrets-manager/service-accounts/projects/service-account-projects.component.ts

Move to emitters

- bitwarden_license/bit-web/src/app/secrets-manager/service-accounts/service-accounts.component.ts

Refresh the service account list on access removal

- bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/access-policy.service.ts

Add in handling new server properties and adding in a null check prior to decryption.

- bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/access-selector.component.html
bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/access-selector.component.ts

Swapping to emitter strategy to handle dialog warnings.

- bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/dialogs/access-removal-dialog.component.html
bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/dialogs/access-removal-dialog.component.ts

Smart dialog for handling access removals when the user would lose access to the target object. 

### Removing user access from service account
image

### Removing group access from service account 
image

### `Can read, write` → `Can read` for user's group project access
image

### Remove user's group project access
image

### Remove user's project access
image

### Can read, write` → `Can read` for user's project access

image

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/dialogs/access-removal-dialog.component.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    }
+    this.dialogRef.close();
+  };
+
+  cancel = () => {
+    if (this.data.type == "project") {
+      this.accessPolicyService.refreshProjectAccessPolicyChanges();
+    } else if (this.data.type == "service-account") {
+      this.accessPolicyService.refreshServiceAccountAccessPolicyChanges();
+    }
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/dialogs/access-removal-dialog.component.ts:62]: Why do we need to refresh these?

File: bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/dialogs/access-removal-dialog.component.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  type: "project" | "service-account";
+  returnRoute: string[];
+  policy: AccessSelectorRowView;
+}
+
+@Component({
+  selector: "sm-access-removal-dialog",
+  templateUrl: "./access-removal-dialog.component.html",
+})
+export class AccessRemovalDialogComponent implements OnInit {
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/dialogs/access-removal-dialog.component.ts:21]: I somewhat wonder if we should just use the configurable dialog here instead. While manually writing the dialog does give us the ability to show a progress bar on remove, we already show process on the regular page.

Thoughts?

File: bitwarden_license/bit-web/src/app/secrets-manager/service-accounts/people/service-account-people.component.ts [hunk 3]
```diff
@@ -92,16 +113,82 @@ export class ServiceAccountPeopleComponent {
...
+      this.rows = rows;
     });
   }
 
   ngOnDestroy(): void {
     this.destroy$.next();
     this.destroy$.complete();
   }
+
+  private async needToShowWarning(
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/secrets-manager/service-accounts/people/service-account-people.component.ts:158]: This seems duplicated between the components?

**Review by Hinton — COMMENTED** [review]
There is some duplicated code between two components, any way we could reuse that?

File: bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/dialogs/access-removal-dialog.component.ts (hunk 1 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/dialogs/access-removal-dialog.component.ts:62]: @Hinton 

With how the  `<select bitInput` is working in the access-selector, the `value` of the select box is already changed when all this gets triggered. To cancel the operation, we need to refresh the access list to display the correct values.

File: bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/dialogs/access-removal-dialog.component.ts (hunk 2 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/secrets-manager/shared/access-policies/dialogs/access-removal-dialog.component.ts:21]: As long as we need to refresh to preform the cancelation, need this to be a regular dialog.

File: bitwarden_license/bit-web/src/app/secrets-manager/service-accounts/people/service-account-people.component.ts (hunk 3 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/secrets-manager/service-accounts/people/service-account-people.component.ts]: Cleaned up as much duplication as I could https://github.com/bitwarden/clients/pull/4904/commits/10b606f42716b4058a0c6fe93c598e5e6af039df

**Review by willmartian — DISMISSED** [review]
Good work!

**Review by willmartian — APPROVED** [review]

**Pull request merged**